### PR TITLE
Geospatially enabled viewports

### PR DESCRIPTION
**Pull request by ibgreen**

### CODE REVIEW

File: src/viewport-mercator-project/web-mercator-utils.js [hunk 1]
```diff
@@ -22,6 +24,15 @@ function createMat4() {
   return [1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1];
 }
 
+// Returns the zoom level that gives a 1 meter pixel at a certain latitude
+// S=C*cos(y)/2^(z+8)
+export function getMeterZoom({latitude}) {
+  assert(latitude);
+  const EARTH_CIRCUMFERENCE = 40.075e6;
```
Comment (Pessimistress) [inline src/viewport-mercator-project/web-mercator-utils.js]: Define at top of file with other constants?

File: src/viewports/viewport.js [hunk 2]
```diff
@@ -39,8 +54,29 @@ const DEFAULT_DISTANCE_SCALES = {
   degreesPerPixel: [1, 1, 1]
 };
 
+const DEFAULT_ZOOM = 0;
+
 const ERR_ARGUMENT = 'Illegal argument to Viewport';
 
+// Extracts or creates projection matrix from supplied options
+// Optionally creates a perspective or orthographic matrprojectionMatrix || ix
+function createPerspectiveProjectionMatrix({ // viewport arguments
```
Comment (Pessimistress) [inline src/viewports/viewport.js]: Should this be part of `math.gl`? Similar to [gl-matrix/mat4.perspective](http://glmatrix.net/docs/module-mat4.html#.perspective)

File: src/viewports/viewport.js [hunk 3]
```diff
@@ -49,33 +85,121 @@ export default class Viewport {
...
+    // Calculate distance scales if lng/lat/zoom are provided
+    this.distanceScales = this.isGeospatial ?
+      getMercatorDistanceScales({latitude, longitude, scale: this.scale}) :
+      distanceScales || DEFAULT_DISTANCE_SCALES;
+
+    this.focalDistance = opts.focalDistance || 1;
+
+    const SCALE = 1;
+
+    this.distanceScales.metersPerPixel =
```
Comment (Pessimistress) [inline src/viewports/viewport.js]: Does this do anything with `SCALE` as a constant?

File: src/viewports/viewport.js [hunk 4]
```diff
@@ -266,20 +433,8 @@ export default class Viewport {
 
     this.pixelUnprojectionMatrix = mat4_invert(createMat4(), this.pixelProjectionMatrix);
     if (!this.pixelUnprojectionMatrix) {
-      throw new Error('Pixel project matrix not invertible');
+      log.warn('Pixel project matrix not invertible');
```
Comment (Pessimistress) [inline src/viewports/viewport.js]: 👍

File: src/viewports/viewport.js (hunk 3 above)
Comment (ibgreen) [reply src/viewports/viewport.js]: > Does this do anything with SCALE as a constant?

Nope, removed.

File: src/viewports/viewport.js (hunk 2 above)
Comment (ibgreen) [reply src/viewports/viewport.js]: > Should this be part of math.gl? Similar to gl-matrix/mat4.perspective

Yes, removed. This did more initially but ended up being essentially the same function as Matrix4().perspective.

**Pull request merged**